### ISSUE TITLE
building with gcc 4.4

### ISSUE THREAD

**Issue by romw** [BOINC/boinc-ng#831]
**Reported by mjakubicek on 22 Sep 39162311 02:40 UTC**
Here are two patches to make BOINC build with gcc 4.4 -- the first one is sane, just simple things, while the other one comes from a colleague and...works, but I don't know why. To be more precise: I don't know why it does not build without it.

**Comment by romw** [comment]
**Commented by romw on 1 May 39173289 20:26 UTC**
The main path here applies to the server code.

**Comment by romw** [comment]
**Commented by davea on 7 Sep 39175738 08:00 UTC**
(In [17502]) - client: don't allow coproc apps in app_info.xml.
    Otherwise we'll get stuck in a loop where the client asks for CPU work,
    and the scheduler sends jobs for what it thinks is a CPU app
    but is actually a coproc app.

    Eventually we should add coproc info to the app descriptions
    send in scheduler request,
    so that you can use anonymous platform for coproc apps.
    But let's wait on this.

- scheduler: compile fix for gcc 4.4.  Fixes #854

**Comment by romw** [comment]
**Commented by romw on 2 Apr 39180788 11:06 UTC**
(In [17527])     - client: don't allow coproc apps in app_info.xml.
        Otherwise we'll get stuck in a loop where the client asks for CPU work,
        and the scheduler sends jobs for what it thinks is a CPU app
        but is actually a coproc app.

        Eventually we should add coproc info to the app descriptions
        send in scheduler request,
        so that you can use anonymous platform for coproc apps.
        But let's wait on this.

    - scheduler: compile fix for gcc 4.4.  Fixes #854

    client/
        cs_statefile.cpp
    sched/
        sched_driver.cpp
        sched_util.cpp

**Comment by romw** [comment]
**Commented by mjakubicek on 26 Aug 39550061 17:46 UTC**
Please fix this also in the 6.6 branch, thanks in advance.

**Comment by romw** [comment]
**Commented by Ageless on 22 Apr 39551052 10:40 UTC**
It '''is''' in the 6.6a branch. See [17527]

**Comment by romw** [comment]
**Commented by mjakubicek on 20 Nov 39553285 01:20 UTC**
Confirm, thanks for pointing to the 6.6a branch, please close this again anyone with appropriate rights.

**Comment by romw** [comment]
**Commented by mjakubicek on 13 Nov 39553489 08:53 UTC**
Ops, sorry, I just found out that the patch I attached in February is outdated, there is one more thing that needs to be resolved:

Starting with glibc 2.10, functions that take const arguments return const values too. This was not the case before, now e.g. strcasestr returns "const char*", not "char*" when the supplied arguments are "const char*". This breaks the build on the following place:

```
>g++ -DHAVE_CONFIG_H -I. -I..  -I../lib -I../api -I../db -I../client -I../tools -I../sched -I../lib/mac -pthread   -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -msse2 -march=opteron -mtune=generic -MT boinc_client-cpu_sched.o -MD -MP -MF .deps/boinc_client-cpu_sched.Tpo -c -o boinc_client-cpu_sched.o `test -f 'cpu_sched.cpp' || echo './'`cpu_sched.cpp
In file included from cpu_sched.cpp:46:
../lib/str_util.h:41: error: new declaration char* strcasestr(const char*, const char*)
/usr/include/string.h:367: error: ambiguates old declaration const char* strcasestr(const char*, const char*)
```

There are indeed two issues to be resolved, first, the fallback declaration should be consistent with that one of current glibc, hence it should be in lib/str_util.h:

```extern const char *strcasestr(const char *s1, const char *s2);```

instead of:

```extern char *strcasestr(const char *s1, const char *s2);```

Moreover, lib/str_util.h is missing #include "config.h" which means that the "HAVE_*" macros found out by configure are never defined (and therefore the conflicting declaration comes out).
I've attached a small patch which resolves both these issues.

**Comment by romw** [comment]
**Commented by Nicolas on 29 Jan 39553514 23:06 UTC**
Public headers should not include `config.h`, and `str_util.h` is "public" for science apps. Code in trunk defines `strcasestr` in a new file that includes `config.h`. Your patch that includes config.h from str_util.h would not be applied.

**Comment by romw** [comment]
**Commented by davea on 29 Dec 39553534 20:26 UTC**
(In [18633]) - lib: gcc 4.4 fix; fixes #854
